pkg/operator/events/recorder.go

```diff
@@ -0,0 +1,92 @@
+package events
+
+import (
+	"fmt"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type RecorderSourceFunc func() (RecorderSource, error)
+
+type Recorder interface {
+	CreateEvent(reason, message string) error
+	CreateEventWarning(reason, message string) error
+}
+
+type RecorderSource interface {
+	GroupVersionKind() schema.GroupVersionKind
+	metav1.ObjectMetaAccessor
+}
+
+func NewRecorder(client corev1client.EventInterface, sourceComponent string, sourceFunc RecorderSourceFunc) Recorder {
+	return &recorder{
+		eventClient:     client,
+		sourceFunc:      sourceFunc,
+		sourceComponent: sourceComponent,
+	}
+}
+
+type recorderSourceObject struct {
+	metav1.TypeMeta
+	metav1.ObjectMetaAccessor
+}
+
+type recorder struct {
+	eventClient     corev1client.EventInterface
+	sourceFunc      RecorderSourceFunc
+	sourceComponent string
+}
+
+func (r *recorder) CreateEvent(reason, message string) error {
+	source, err := r.sourceFunc()
+	if err != nil {
+		return fmt.Errorf("enable to get event source: %v", err)
+	}
+	event := r.makeEvent(source, "Normal", reason, message)
+	_, err = r.eventClient.Create(event)
+	return err
+}
+
+func (r *recorder) CreateEventWarning(reason, message string) error {
+	source, err := r.sourceFunc()
+	if err != nil {
+		return fmt.Errorf("enable to get event source: %v", err)
+	}
+	event := r.makeEvent(source, "Warning", reason, message)
+	_, err = r.eventClient.Create(event)
+	return err
+}
+
+func (r recorder) makeEvent(source RecorderSource, eventType, reason, message string) *corev1.Event {
+	currentTime := metav1.Time{Time: time.Now()}
+	gvk := source.GroupVersionKind()
+	apiVersion, kind := gvk.ToAPIVersionAndKind()
+
+	event := &corev1.Event{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%v.%x", source.GetObjectMeta().GetName(), currentTime.UnixNano()),
```

Comment:
never looked into how events are named. this is how? 😱

Reply:
yes, copied from deployment controller :)